File: leetcode/Spiral_Matrix_II.java
import java.util.*;

public class Solution {
    void print(int[][] m, int b, int e, int k) {
        if (b == e) {
            m[b][b] = k;
            return;
        }
        for (int j = b; j < e; ++j) m[b][j] = k++;
        for (int i = b; i < e; ++i) m[i][e] = k++;
        for (int j = e; j > b; --j) m[e][j] = k++;
        for (int i = e; i > b; --i) m[i][b] = k++;
    }
    public int[][] generateMatrix(int n) {
        int[][] m = new int[n][n];
        int k = 1;
        for (int i = 0; i <= n - 1 - i; ++i) {
            print(m, i, n - 1 - i, k);
            k += (n - 2 * i) * 4 - 4;
        }
        return m;
    }
}

File: leetcode/Spiral_Matrix_II.cpp
int dir[4][2] = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};

class Solution {
public:
    vector<vector<int> > generateMatrix(int n) {
        vector<vector<int> > matrix(n, vector<int>(n, 0));
        int x = 0;
        int y = 0;
        int k = 0;  // direction
        for (int i = 1; i <= n * n; ++i) {
            matrix[x][y] = i;
            int nx = x + dir[k][0];
            int ny = y + dir[k][1];
            if (nx >= 0 && nx < n && ny >= 0 && ny < n && matrix[nx][ny] == 0) {
                x = nx;
                y = ny;
            } else {
                k = (k + 1) % 4;
                x += dir[k][0];
                y += dir[k][1];
            }
        }
        return matrix;
    }
};
